portfolio/src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -18,44 +18,107 @@
 import com.google.appengine.api.users.UserService;
 import com.google.appengine.api.users.UserServiceFactory;
 import com.google.gson.Gson;
+import com.google.gson.JsonObject;
 import java.io.IOException;
 import java.io.PrintWriter;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-@WebServlet("/Authenticate")
+@WebServlet("/user-data")
 public class LoginServlet extends HttpServlet {
 
+    // doGet is used by comment sumbission area.
+    // if user is logged in: Provides user-data for sending comments.
+    // if user is logged out: provides login link.
     @Override
     public void doGet (HttpServletRequest request, HttpServletResponse response) throws IOException {
         response.setContentType("application/json");
         UserService userService = UserServiceFactory.getUserService();
-        String userId = userService.getCurrentUser().getUserId();
 
-        String nickname = getUserNickname(userId);
-        if (nickname == null) nickname = "Anonymous";
+        JsonObject json = new JsonObject();
 
-        response.getWriter().println("{ \"nickname\" : \"" + nickname + "\" }");
+        if (!userService.isUserLoggedIn()) {
+            String loginUrl = userService.createLoginURL("/");
+            json.addProperty("loginUrl", loginUrl);
+            json.addProperty("userLoggedIn", false);
+        } else {
+            String nickname = getCurrentUserNickname();
+            if (nickname == null) {
```

Comment:
It looks like there's a couple places that you allow a null nickname to be replaced with a default value. While this technically works, I'd encourage you to consider if that's actually the behavior you want. In a lot of cases, catching and ignoring errors like this can make your program hard to debug in the future. If you really do want a null nickname to be ok, leave a comment explaining why that's the case. Otherwise, update these cases to print an error, or otherwise indicate that something unexpected went wrong.